### PR TITLE
feat: support write data into special retention policy

### PR DESCRIPTION
**Pull request by Chenxulin97**
Feat #80

### CODE REVIEW

**Review by shoothzj — CHANGES_REQUESTED** [review]
- Let's use `WithRp` to keep consistent
- We need to update the sdk design doc, it's been merged in https://github.com/openGemini/openGemini.github.io/pull/109 https://github.com/openGemini/openGemini.github.io/pull/108

**Comment by shoothzj** [comment]
IIUC, `Feat #80` will not close any issue, it's invalid

**Comment by Chenxulin97** [comment]
@shoothzj  I add the write with rp design in this pr [openGemini.github.io#111](https://github.com/openGemini/openGemini.github.io/pull/111/files)

File: opengemini/client.go [hunk 1]
```diff
@@ -23,6 +23,11 @@ type Client interface {
 	WritePoint(ctx context.Context, database string, point *Point, callbackFunc WriteCallback) error
 	// WriteBatchPoints write batch points to assigned database
 	WriteBatchPoints(database string, bp []*Point) error
+	// WritePointWithRp write single point into special retention policy and assigned database
+	WritePointWithRp(ctx context.Context, database string, rp string, point *Point, callbackFunc WriteCallback) error
```
Comment (shoothzj) [inline opengemini/client.go:27]: can we keep order consitent with design doc?

Comment (Chenxulin97) [reply opengemini/client.go:25]: done

File: opengemini/client.go [hunk 2]
```diff
@@ -21,8 +21,13 @@ type Client interface {
 
 	// WritePoint write single point to assigned database
 	WritePoint(ctx context.Context, database string, point *Point, callbackFunc WriteCallback) error
+	// WritePointWithRp write single point into special retention policy and assigned database
+	WritePointWithRp(ctx context.Context, database string, rp string, point *Point, callbackFunc WriteCallback) error
```
Comment (xuthus5) [inline opengemini/client.go:25]: give a shorter definition, x, x type

File: opengemini/write.go [hunk 3]
```diff
@@ -105,7 +165,38 @@ func (c *client) WritePoint(ctx context.Context, database string, point *Point,
 	return nil
 }
 
-func (c *client) internalBatchSend(ctx context.Context, database string, resource <-chan *sendBatchWithCB) {
+func (c *client) WriteBatchPointsWithRp(database string, rp string, bp []*Point) error {
```
Comment (xuthus5) [inline opengemini/write.go:168]: duplicate code, the same with `WriteBatchPoints` , try to make it simpler

**Review by cyruslo — APPROVED** [review]

File: opengemini/write.go [hunk 4]
```diff
@@ -50,16 +50,76 @@ type sendBatchWithCB struct {
...
+			return errors.New("writer close failed, error: " + err.Error())
+		}
+	}
+
+	resp, err := c.innerWrite(database, "", &buffer)
+	if err != nil {
+		callback(errors.New("innerWrite request failed, error: " + err.Error()))
+	}
+
+	defer resp.Body.Close()
```
Comment (cyruslo) [inline opengemini/write.go:98]: 记得判空，95行没返回

File: opengemini/write.go (hunk 3 above)
Comment (Chenxulin97) [reply opengemini/write.go:168]: accept

File: opengemini/write.go (hunk 4 above)
Comment (Chenxulin97) [reply opengemini/write.go:98]: accept

**Review by shoothzj — APPROVED** [review]

**Comment by shoothzj** [comment]
LGTM

**Pull request merged**